text/0939-import-glob.md

```diff
@@ -0,0 +1,196 @@
+> Ember apps are authored as JavaScript Modules (also know as "ES Modules"). By convention, your app's modules live in the `/app` directory, so if you see an import like `import Article from "your-app/models/article"`, that is referring to `/app/models/article.js`. If you install dependencies from NPM you can import from them as well.
+> 
+> In a default Ember app, you can use dynamic `import()` to load third-party modules from NPM on demand, but you can't use it on your own app code. (That feature is available if you're building with [Embroider](https://github.com/embroider-build/embroider/), but that is not the default experience yet.)
+>
+> If you need to import many modules at once, Ember apps support an extension on top of ES modules called `import.meta.glob()`. For example, `import.meta.glob('./widgets/*.js')` will give you access to all the matching files. You can only use `import.meta.glob()` on files within your own package.
+
+The more detailed nuances of what `import.meta.glob` supports should be taught through good error messages. For example, all of these need to give clear explanation in an error messages:
+
+ - trying to pass a non-string-literal to `import.meta.glob`
+ - trying to escape your package via `../`
+ - trying to import a pattern that doesn't start with `./` or `../`
+
+## Drawbacks
+
+As designed, this is not a drop-in replacement for the old system. Addons that relied on the looseness of the old system are going to need to make breaking changes to their public API to adapt to this change. I think those breaking changes are likely to be "constant cost" changes that are not expensive, even for big application, so I consider this worth it in order to get us into a more long-term-supportable position that is compatible with standard Javascript.
+
+## Alternatives
+
+### Globally powerful import.meta.glob
+
+We could attempt to allow more globally-powerful `import.meta.glob`. For example, it might be possible to make patterns starting with `/` always search the current application, even when an addon is asking. This would give addon authors more of the power they're used to having, but I think it's a much riskier feature to enable across the ecosystem. I'm not convinced we could make it work at reasonable cost in even all current build systems, never mind future ones. As written, this RFC has low-risk of causing compatibility problems in the future since the feature is not allowed in addon publication format. This makes it much easier to evolve the feature over time without breaking the universe.
+
+### Additional Vite feature space
+
+Features from Vite's implementation that I didn't incorporate because I don't want to sign us up to reimplement them in every build system:
+
+ - absolute imports, starting with `/`. This is not a well-defined concept in today's Ember apps because they do not have a single directory representing the app's web root.
+ - eager mode, so that you gain synchronous access to the matching modules
```

Comment:
I think I worked this out between the two RFCs, but wanted to keep my thought process here in case others have the same/similar thoughts

----

I've used eager in my spikes of wiring up Qunit in Vite.

since we're using native modules though, and native modules support top-level await, we could effectively re-implement the behavior, if anyone needs(?) this way:
```js
import * as QUnit from 'qunit';
import { setup } from 'qunit-dom';

setup(QUnit.assert);

const moduleRequests = import.meta.glob('./**/*test.{js,ts}');

// loads all the test files, which are then registered with QUnit
await Promise.all(Object.values(moduleRequests).map(loader => loader()));
```

Though, this requires that our distribution format is ESM, yeah? like, we couldn't keep shipping AMD if we wanted to do something like this.
(

I think that's the point of the other RFC tho?: https://github.com/emberjs/rfcs/pull/938

in particular,

>  - we cannot implement other standard ECMA features like top-level await until these incompatible APIs are gone (consider: how is `require()` supposed to deal with discovering a transitive dependency that contains top-level await?)

)

Reply:
I don't want to create a hard dependency between this change and top-level await, so I don't think we should say that's the upgrade path. We want to implement `import.meta.glob` in classic builds too, where we may never support top-level await.

But we don't need top-level await to achieve a compatible way to wire up tests. Both because (1) Embroider already synthesizes a test entrypoint that works in vite, and (2) even if you wanted to redo that to use import.meta.glob, it's up to us when to start the qunit tests. 

```js
QUnit.config.autostart = false;
Promise.all(Object.values(moduleRequests).map(loader => loader())).then(() => QUnit.start());
```

I do think your question points toward further simplifications of our programming model though. It would be possible to codify the "what is an ember app and how do you boot it" in terms of import.meta.glob. Today the conventions are implicit and Embroider and classic both try to implement them by having opinions about what files to stitch together into the entrypoint of an application. But we could retcon that in terms of import.meta.glob and make it visible to users.

This would be for a v2 app RFC and/or Polaris Routing RFC. But you can imagine how it would work:

```js
const MyRouter = Router.map(import.meta.glob("./routes/**/*.ts", { ... })
```

The examples @wycats showed at EmberConf were similar but they said `Router.map(import.meta, { ... })`, and I've confirmed that's not really going to work in existing build tools like Vite. They have the static requirement on import.meta.glob that I've encoded in this RFC. Passing the *output* of import.meta.glob though would work, and would make it clear which "entry points" you're telling the app to work with.